File: glTest/src/glTest/framework/ApplicationState.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package glTest.framework;

import com.jogamp.newt.Display;
import com.jogamp.newt.NewtFactory;
import com.jogamp.newt.Screen;
import com.jogamp.newt.event.KeyEvent;
import com.jogamp.newt.event.KeyListener;
import com.jogamp.newt.opengl.GLWindow;
import com.jogamp.opengl.FPSCounter;
import static com.jogamp.opengl.GL.GL_BLEND;
import static com.jogamp.opengl.GL.GL_CULL_FACE;
import static com.jogamp.opengl.GL.GL_DEPTH_TEST;
import static com.jogamp.opengl.GL.GL_DONT_CARE;
import static com.jogamp.opengl.GL.GL_FRONT;
import static com.jogamp.opengl.GL.GL_LESS;
import static com.jogamp.opengl.GL.GL_RENDERER;
import static com.jogamp.opengl.GL.GL_SCISSOR_TEST;
import static com.jogamp.opengl.GL.GL_VENDOR;
import static com.jogamp.opengl.GL.GL_VERSION;
import static com.jogamp.opengl.GL2ES2.GL_DEBUG_SEVERITY_HIGH;
import static com.jogamp.opengl.GL2ES2.GL_DEBUG_SEVERITY_MEDIUM;
import static com.jogamp.opengl.GL2ES2.GL_SHADING_LANGUAGE_VERSION;
import com.jogamp.opengl.GL4;
import com.jogamp.opengl.GLAutoDrawable;
import com.jogamp.opengl.GLCapabilities;
import com.jogamp.opengl.GLContext;
import com.jogamp.opengl.GLEventListener;
import com.jogamp.opengl.GLProfile;
import com.jogamp.opengl.util.Animator;
import com.jogamp.opengl.util.GLBuffers;
import glm.vec._2.i.Vec2i;
import java.util.ArrayList;
import glTest.problems.Problem;
import glTest.solutions.Solution;
import java.nio.IntBuffer;

/**
 *
 * @author GBarbieri
 */
public class ApplicationState implements GLEventListener, KeyListener {

    public static final Vec2i RESOLUTION = new Vec2i(1024, 768);
    private ProblemFactory factory;
    private ArrayList<Problem> problems;
    private Solution[] solutions;
    private Problem problem;
    private Solution solution;
    public GLWindow glWindow;
    public static Animator animator;
    public static boolean DEBUG = true;
    private IntBuffer vertexArrayObject = GLBuffers.newDirectIntBuffer(1);
    private final String rootTitle = "gltest";
    private int offsetProblem = 0;
    private int offsetSolution = 0;

    public void setup() {

        Display display = NewtFactory.createDisplay(null);
        Screen screen = NewtFactory.createScreen(display, 0);
        GLProfile glProfile = GLProfile.get(GLProfile.GL4);
        GLCapabilities glCapabilities = new GLCapabilities(glProfile);
        glWindow = GLWindow.create(screen, glCapabilities);

        glWindow.setSize(1024, 768);
        glWindow.setPosition(50, 50);
        glWindow.setUndecorated(false);
        glWindow.setAlwaysOnTop(false);
        glWindow.setFullscreen(false);
        glWindow.setPointerVisible(true);
        glWindow.confinePointer(false);
        glWindow.setTitle(rootTitle);
        if (DEBUG) {
            glWindow.setContextCreationFlags(GLContext.CTX_OPTION_DEBUG);
        }

        glWindow.setVisible(true);

        if (DEBUG) {
            glWindow.getContext().addGLDebugListener(new GlDebugOutput());
        }

        glWindow.addGLEventListener(this);
        glWindow.addKeyListener(this);

        System.out.println("GL created successfully! Info follows.");

        animator = new Animator(glWindow);
        animator.setRunAsFastAsPossible(true);
        animator.setUpdateFPSFrames(FPSCounter.DEFAULT_FRAMES_PER_INTERVAL, System.out);
        animator.start();
    }

    @Override
    public void init(GLAutoDrawable drawable) {

        GL4 gl4 = drawable.getGL().getGL4();
        System.out.println("" + GLContext.getCurrent().getGLVersion());
        System.out.println("Vendor: " + gl4.glGetString(GL_VENDOR));
        System.out.println("Renderer: " + gl4.glGetString(GL_RENDERER));
        System.out.println("Version: " + gl4.glGetString(GL_VERSION));
        System.out.println("Shading Language Version: " + gl4.glGetString(GL_SHADING_LANGUAGE_VERSION));

        gl4.setSwapInterval(0);

        // Default GL State
        gl4.glCullFace(GL_FRONT);
        gl4.glEnable(GL_CULL_FACE);
        gl4.glDisable(GL_SCISSOR_TEST);
        gl4.glEnable(GL_DEPTH_TEST);
        gl4.glDepthMask(true);
        gl4.glDepthFunc(GL_LESS);
        gl4.glDisable(GL_BLEND);
        gl4.glColorMask(true, true, true, true);

        if (DEBUG) {
            gl4.glDebugMessageControl(GL_DONT_CARE, GL_DONT_CARE, GL_DONT_CARE, 0, null, false);
            gl4.glDebugMessageControl(GL_DONT_CARE, GL_DONT_CARE, GL_DEBUG_SEVERITY_HIGH, 0, null, true);
            gl4.glDebugMessageControl(GL_DONT_CARE, GL_DONT_CARE, GL_DEBUG_SEVERITY_MEDIUM, 0, null, true);
        }

        // Now that we have something valid, create our VAO and bind it. Ugh! So lame that this is required.
        gl4.glGenVertexArrays(1, vertexArrayObject);
        gl4.glBindVertexArray(vertexArrayObject.get(0));

        factory = new ProblemFactory();
        problems = factory.getProblems();
        assert (problems.size() > 0);

        setInitialProblemAndSolution(gl4, "NullProblem", "NullSolution");
    }

    private void setInitialProblemAndSolution(GL4 gl4, String probName, String solnName) {

        for (int i = 0; i < problems.size(); i++) {
            if (problems.get(i).getName().equals(probName)) {
                problem = problems.get(i);
                break;
            }
        }

        solutions = factory.getSolutions(problem);
        for (Solution sol : solutions) {
            if (sol.getName().equals(solnName)) {
                solution = sol;
                break;
            }
        }

        initProblem(gl4);

        initSolution(gl4, 0);

        onProblemOrSolutionSet();
    }

    @Override
    public void display(GLAutoDrawable drawable) {

        GL4 gl4 = drawable.getGL().getGL4();

        if (offsetProblem != 0) {
            changeProblem(gl4);
        }
        if (offsetSolution != 0) {
            changeSolution(gl4);
        }

        if (problem == null) {
            return;
        }

        // This is the main entry point shared by all tests. 
        problem.render(gl4);

        // Present the results.
        // included in the animator
    }

    @Override
    public void reshape(GLAutoDrawable drawable, int x, int y, int width, int height) {

        GL4 gl4 = drawable.getGL().getGL4();

        gl4.glViewport(0, 0, width, height);
    }

    @Override
    public void dispose(GLAutoDrawable drawable) {

        GL4 gl4 = drawable.getGL().getGL4();

        // Must cleanup before we call base class.
        gl4.glBindVertexArray(0);
        gl4.glDeleteVertexArrays(1, vertexArrayObject);

        System.exit(0);
    }

    @Override
    public void keyPressed(KeyEvent e) {
        switch (e.getKeyCode()) {
            case KeyEvent.VK_ESCAPE:
                animator.stop();
                break;
            case KeyEvent.VK_LEFT:
                offsetProblem = -1;
                break;
            case KeyEvent.VK_RIGHT:
                offsetProblem = 1;
                break;
            case KeyEvent.VK_UP:
                offsetSolution = -1;
                break;
            case KeyEvent.VK_DOWN:
                offsetSolution = 1;
                break;
        }
    }

    @Override
    public void keyReleased(KeyEvent e) {

    }

    private void changeProblem(GL4 gl4) {

        shutdownSolution(gl4);

        shutdownProblem(gl4);

        int problemCount = problems.size();
        int problemId = problems.indexOf(problem);
        problemId = (problemId + problemCount + offsetProblem) % problemCount;

        problem = problems.get(problemId);

        initProblem(gl4);

        solutions = factory.getSolutions(problem);

        solution = solutions[problem.getSolutionId()];

        initSolution(gl4, problem.getSolutionId());

        offsetProblem = 0;

        onProblemOrSolutionSet();
    }

    private void changeSolution(GL4 gl4) {

        shutdownSolution(gl4);

        int solutionCount = solutions.length;
        if (solutionCount == 0) {
            return;
        }

        int solutionId = problem.getSolutionId();
        solutionId = (solutionId + solutionCount + offsetSolution) % solutionCount;
        
        solution = solutions[solutionId];

        initSolution(gl4, solutionId);

        offsetSolution = 0;

        onProblemOrSolutionSet();
    }

    private void initSolution(GL4 gl4, int solutionId) {

        System.out.print("Solution " + solution.getName() + " init... ");
        System.out.println(solution.init(gl4) ? "Ok" : "Fail");

        problem.setSolution(gl4, solution);
        problem.setSolutionId(solutionId);
    }

    private void shutdownSolution(GL4 gl4) {

        System.out.print("Solution " + solution.getName() + " shutdown... ");
        System.out.println(solution.shutdown(gl4) ? "Ok" : "Fail");

        problem.setSolution(gl4, null);
    }

    private void initProblem(GL4 gl4) {

        System.out.print("Problem " + problem.getName() + " - init... ");
        System.out.println(problem.init(gl4) ? "Ok" : "Fail");
    }

    private void shutdownProblem(GL4 gl4) {

        System.out.print("Problem " + problem.getName() + " shutdown... ");
        System.out.println(problem.shutdown(gl4) ? "Ok" : "Fail");
    }

    private void onProblemOrSolutionSet() {

        System.gc();
        
        String newTitle = rootTitle + " - " + problem.getName();

        if (solution != null) {
            newTitle += " - " + solution.getName();
        }
        glWindow.setTitle(newTitle);
        
        System.gc();

        animator.resetFPSCounter();
    }
}

File: src/main/kotlin/main/glTest/framework/ApplicationState.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package main.glTest.framework

import java.util.ArrayList
import glm_.vec2.Vec2i
import org.lwjgl.system.Platform
import uno.glfw.glfw
import java.lang.Error

fun main(){
    ApplicationState()
}

const val DEBUG = true

const val inactiveProblem = -1
const val inactiveSolution = -1

object Options {

    var benchmarkMode = false
    var benchmarkTime = 5f
    val resolution = Vec2i(1024, 768)

    val defaultInitialProblem = "NullProblem"
    val defaultInitialSolution = "NullSolution"

    var initialProblem = defaultInitialProblem
    var initialSolution = defaultInitialSolution

    var initialApi = when(Platform.get()) {
        Platform.MACOSX -> OpenGLCore.shortName
        else -> OpenGLGeneric.shortName
    }
}

/**
 *
 * @author GBarbieri
 */
class ApplicationState {

    init {
        glfw.init()
    }

    val resolution = Options.resolution

    val activeProblem = inactiveProblem
    val activeSolution = inactiveSolution

    val glApis = mutableMapOf<String, OpenGLBase>()

    // ~ createGfxApis
    init {
        val generic = OpenGLGeneric()
        if (generic.init("apitest - OpenGL (compat)", Vec2i(), resolution))
            glApis[OpenGLGeneric.shortName] = generic

        val core = OpenGLCore()
        if (core.init("apitest - OpenGL (core)", Vec2i(), resolution))
            glApis[OpenGLCore.shortName] = core
    }

    val activeGl = glApis[Options.initialApi] ?: throw Error("Failed to select api with name '${Options.initialApi}'")

    init {
        activeGl.activate()
    }

    val factory = ProblemFactory()
    private var problems: ArrayList<Problem>? = null
    private var solutions: Array<Solution>? = null
    private var problem: Problem? = null
    private var solution: Solution? = null
    private val vertexArrayName = GLBuffers.newDirectIntBuffer(1)
    private val queryName = GLBuffers.newDirectIntBuffer(1)
    private val gpuTime = GLBuffers.newDirectLongBuffer(1)
    private var offsetProblem = 0
    private var offsetSolution = 0
    private var frames = 0
    private var cpuStart: Long = 0
    private var cpuTotal: Long = 0
    private var gpuTotal: Long = 0
    private var updateCountersStart: Long = 0
    private val updateTick: Long = 1000

    fun init(drawable: GLAutoDrawable) {

        val gl4 = drawable.getGL().getGL4()
        System.out.println("" + GLContext.getCurrent().getGLVersion())
        System.out.println("Vendor: " + gl4.glGetString(GL_VENDOR))
        System.out.println("Renderer: " + gl4.glGetString(GL_RENDERER))
        System.out.println("Version: " + gl4.glGetString(GL_VERSION))
        System.out.println("Shading Language Version: " + gl4.glGetString(GL_SHADING_LANGUAGE_VERSION))

        gl4.setSwapInterval(0)

        // Default GL State
        gl4.glCullFace(GL_FRONT)
        gl4.glEnable(GL_CULL_FACE)
        gl4.glDisable(GL_SCISSOR_TEST)
        gl4.glEnable(GL_DEPTH_TEST)
        gl4.glDepthMask(true)
        gl4.glDepthFunc(GL_LESS)
        gl4.glDisable(GL_BLEND)
        gl4.glColorMask(true, true, true, true)

        if (DEBUG) {
            gl4.glDebugMessageControl(GL_DONT_CARE, GL_DONT_CARE, GL_DONT_CARE, 0, null, false)
            gl4.glDebugMessageControl(GL_DONT_CARE, GL_DONT_CARE, GL_DEBUG_SEVERITY_HIGH, 0, null, true)
            gl4.glDebugMessageControl(GL_DONT_CARE, GL_DONT_CARE, GL_DEBUG_SEVERITY_MEDIUM, 0, null, true)
        }

        gl4.glGetQueryiv(GL_TIME_ELAPSED, GL_QUERY_COUNTER_BITS, queryName)
        println("GL_QUERY_COUNTER_BITS: " + queryName.get(0))

        gl4.glGenQueries(1, queryName)

        // Now that we have something valid, create our VAO and bind it. Ugh! So lame that this is required.
        gl4.glGenVertexArrays(1, vertexArrayName)
        gl4.glBindVertexArray(vertexArrayName.get(0))

        factory = ProblemFactory()
        problems = factory!!.getProblems()
        assert(problems!!.size > 0)

        setInitialProblemAndSolution(gl4, "NullProblem", "NullSolution")

        updateCountersStart = System.currentTimeMillis()
    }

    private fun setInitialProblemAndSolution(gl4: GL4, probName: String, solnName: String) {

        for (i in problems!!.indices) {
            if (problems!![i].getName() == probName) {
                problem = problems!![i]
                break
            }
        }

        solutions = factory!!.getSolutions(problem)
        for (sol in solutions!!) {
            if (sol.getName() == solnName) {
                solution = sol
                break
            }
        }

        initProblem(gl4)

        initSolution(gl4, 0)

        onProblemOrSolutionSet()
    }

    fun display(drawable: GLAutoDrawable) {

        val gl4 = drawable.getGL().getGL4()

        if (offsetProblem != 0) {
            changeProblem(gl4)
        }
        if (offsetSolution != 0) {
            changeSolution(gl4)
        }

        if (problem == null) {
            return
        }

        gl4.glBeginQuery(GL_TIME_ELAPSED, queryName.get(0))
        run {
            cpuStart = System.nanoTime()
            run {
                // This is the main entry point shared by all tests.
                problem!!.render(gl4)
                frames++
            }
            cpuTotal += System.nanoTime() - cpuStart
        }
        gl4.glEndQuery(GL_TIME_ELAPSED)
        gl4.glGetQueryObjectui64v(queryName.get(0), GL_QUERY_RESULT, gpuTime)
        gpuTotal += gpuTime.get(0)

        // Present the results.
        //        if (frames == problem.getSolution().updateFps) {
        if (System.currentTimeMillis() - updateCountersStart > updateTick) {
            //            System.out.println("cpuTotal: " + cpuTotal / 1_000_000 + ", gpuTotal: " + gpuTotal + ", frames: " + frames);
            val cpu = String.format("%,.3f", cpuTotal.toDouble() / 1000000.0 / frames.toDouble()) + " ms"
            val gpu = String.format("%,.3f", gpuTotal.toDouble() / 1000000.0 / frames.toDouble()) + " ms"
            val fps = String.format("%,.2f", 1000 / (gpuTotal.toDouble() / 1000000.0 / frames.toDouble()))
            //            String fps = String.format("%,.2f", frames / ((double) gpuTotal / 1_000_000));
            println("CPU time: $cpu, GPU time: $gpu, theor. FPS: $fps")
            resetCounter()
        }
    }

    private fun resetCounter() {
        frames = 0
        cpuTotal = 0
        gpuTotal = 0
        updateCountersStart = System.currentTimeMillis()
    }

    fun reshape(drawable: GLAutoDrawable, x: Int, y: Int, width: Int, height: Int) {

        val gl4 = drawable.getGL().getGL4()

        gl4.glViewport(0, 0, width, height)
    }

    fun dispose(drawable: GLAutoDrawable) {

        val gl4 = drawable.getGL().getGL4()

        // Must cleanup before we call base class.
        gl4.glBindVertexArray(0)
        gl4.glDeleteVertexArrays(1, vertexArrayName)

        System.exit(0)
    }

    fun keyPressed(e: KeyEvent) {
        when (e.getKeyCode()) {
            KeyEvent.VK_ESCAPE -> animator.stop()
            KeyEvent.VK_LEFT -> offsetProblem = -1
            KeyEvent.VK_RIGHT -> offsetProblem = 1
            KeyEvent.VK_UP -> offsetSolution = -1
            KeyEvent.VK_DOWN -> offsetSolution = 1
        }
    }

    fun keyReleased(e: KeyEvent) {

    }

    private fun changeProblem(gl4: GL4) {

        shutdownSolution(gl4)

        shutdownProblem(gl4)

        val problemCount = problems!!.size
        var problemId = problems!!.indexOf(problem)
        problemId = (problemId + problemCount + offsetProblem) % problemCount

        problem = problems!![problemId]

        initProblem(gl4)

        solutions = factory!!.getSolutions(problem)

        solution = solutions!![problem!!.getSolutionId()]

        initSolution(gl4, problem!!.getSolutionId())

        offsetProblem = 0

        onProblemOrSolutionSet()
    }

    private fun changeSolution(gl4: GL4) {

        shutdownSolution(gl4)

        val solutionCount = solutions!!.size
        if (solutionCount == 0) {
            return
        }

        var solutionId = problem!!.getSolutionId()
        solutionId = (solutionId + solutionCount + offsetSolution) % solutionCount

        solution = solutions!![solutionId]

        initSolution(gl4, solutionId)

        offsetSolution = 0

        onProblemOrSolutionSet()
    }

    private fun initSolution(gl4: GL4, solutionId: Int) {

        print("Solution " + solution!!.getName() + " init... ")
        println(if (solution!!.init(gl4)) "Ok" else "Fail")

        problem!!.setSolution(gl4, solution)
        problem!!.setSolutionId(solutionId)

        resetCounter()
    }

    private fun shutdownSolution(gl4: GL4) {

        print("Solution " + solution!!.getName() + " shutdown... ")
        println(if (solution!!.shutdown(gl4)) "Ok" else "Fail")

        problem!!.setSolution(gl4, null)
    }

    private fun initProblem(gl4: GL4) {

        print("Problem " + problem!!.getName() + " - init... ")
        println(if (problem!!.init(gl4)) "Ok" else "Fail")
    }

    private fun shutdownProblem(gl4: GL4) {

        print("Problem " + problem!!.getName() + " shutdown... ")
        println(if (problem!!.shutdown(gl4)) "Ok" else "Fail")
    }

    private fun onProblemOrSolutionSet() {

        System.gc()

        var newTitle = rootTitle + " - " + problem!!.getName()

        if (solution != null) {
            newTitle += " - " + solution!!.getName()
        }
        glWindow.setTitle(newTitle)

        System.gc()
    }

    companion object {

        val RESOLUTION = Vec2i(1024, 768)

        @Throws(InterruptedException::class)
        @JvmStatic
        fun main(args: Array<String>) {

            TimeHack6435126.enableHighResolutionTimer()

            val app = ApplicationState()
            app.setup()
        }

        var animator: Animator
    }
}
